toolchain/check/facet_type.cpp

```diff
@@ -433,21 +405,51 @@ class SubstImplWitnessAccessCallbacks : public SubstInstCallbacks {
       }
     }
 
-    auto SetFullyRewritten(Context& context, SemIR::ImplWitnessAccess access,
-                           SemIR::InstId rewritten_to_inst_id) -> void {
+    auto SetFullyRewritten(Context& context, SemIR::LocId loc_id,
+                           SemIR::ImplWitnessAccess access,
+                           SemIR::InstId access_inst_id,
+                           SemIR::InstId rewritten_to_inst_id)
+        -> SemIR::InstId {
       auto it = map_.find({GetKey(context, access)});
       CARBON_CHECK(it != map_.end());
-      // TODO: If state == FullyRewrtten and the inst id is different (according
-      // to `CompareFacetTypeConstraintValues`), we can diagnose writing two
-      // different values for the same associated constant immediately?
-      //
-      // TODO: Once the above is done, we don't need to do the SortAndDedupe
-      // step in ResolveFacetTypeRewriteConstraints()? We can just convert this
-      // `map_` into a new set of `RewriteConstraint`s which will already be
-      // deduped.
-      if (it->second.state == BeingRewritten) {
+      if (it->second.state == FullyRewritten) {
+        auto ord = CompareFacetTypeConstraintValues(context, it->second.inst_id,
```

Comment:
I think this can be simplified further: this is the only caller of `CompareFacetTypeConstraintValues` and only cares about equality, so we can remove the ordering part of it and just do an equality comparison. I think the equality comparison we want would reduce to `context.constant_values().GetConstantInstId(it->second.inst_id) == context.constant_values().GetConstantInstId(rewritten_to_inst_id)`.

Reply:
Why do the `.U` references not evaluate to the same constant value here? (Are the `.Self`s referring to different bindings? Maybe we need to substitute them if so.) Given that your example doesn't work when directly comparing constant values, I would expect that the same thing with `.U` replaced by `C(.U)` would not work in this PR for the same reason.

Reply:
The FacetTypes inside them are different, so they are different constant values. In the innermost, the FacetType of Self is `Z`, in the next one it's `Z where .T = .U`.

Ah good, thanks for the test case.